Implementation of tickets 1604, 1597 and 1365

BatiGencho opened the pull request:
### Description

Implementation of tickets 1604, 1597 and 1365

### Related issues

- Fixes #1604, #1597, #1365 

### Note
Regarding SortedOracles.sol, we are changing the type of values stored in the state variable r"ates" to unwrapped Fixidity fractions meaning those changes are not backwards compatible and shall remain as so onwards. The latter is to take effect on all interacting package levels.

Commit:
added changes to the 3 tickets

jfoutts-celo commented on packages/protocol/contracts/common/MultiSig.sol:
```diff
@@ -117,7 +127,16 @@ contract MultiSig is Initializable {
       isOwner[_owners[i]] = true;
```
Should we emit OwnerAddition?

BatiGencho replied:
added

jfoutts-celo commented on packages/protocol/contracts/governance/EpochRewards.sol:
```diff
@@ -159,6 +161,29 @@ contract EpochRewards is Ownable, Initializable, UsingPrecompiles, UsingRegistry
 
+  /**
+   * @notice Sets the target voting yield for validators.
+   * @param _targetVotingYield The value of the target voting yield.
```
We should document whenever parameters are expected to be unwrapped Fixidity values.

BatiGencho replied:
added

jfoutts-celo commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -344,28 +394,46 @@ contract('SortedOracles', (accounts: string[]) => {
-        })
+        await sortedOracles.report(
+          aToken,
+          new BigNumber(numerator / denominator).multipliedBy(fractionDenominator),
```
Should this be newNumerator?

BatiGencho replied:
fixed

jfoutts-celo commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -115,7 +117,15 @@ contract('SortedOracles', (accounts: string[]) => {
-        await sortedOracles.report(aToken, 1, 1, NULL_ADDRESS, NULL_ADDRESS, { from: anOracle })
+        await sortedOracles.report(
+          aToken,
+          new BigNumber(1 / 1).multipliedBy(fractionDenominator),
```
More clear to remove divisions by one?

jfoutts-celo commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -270,41 +292,64 @@ contract('SortedOracles', (accounts: string[]) => {
-      })
+      await sortedOracles.report(
+        aToken,
+        new BigNumber(numerator / denominator).multipliedBy(fractionDenominator),
```
Nit: `new BigNumber(numerator).dividedBy(new BigNumber(denomninator))` instead of `new BigNumber(numerator / denominator)`.

BatiGencho replied:
doen everywhere

jfoutts-celo commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -127,6 +130,67 @@ contract StableToken is
+
+  /**
+   * @notice Sets initial inflation Parameters.
+   * @param inflationRate new rate.
```
Comment that unwrapped Fixidity value expected for `inflationRate`.

BatiGencho replied:
done

jfoutts-celo commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -121,23 +124,19 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
    * @param token The address of the token for which the Celo Gold exchange rate is being reported.
-   * @param numerator The amount of tokens equal to `denominator` Celo Gold.
-   * @param denominator The amount of Celo Gold equal to `numerator` tokens.
+   * @param value The amount of tokens equal to Celo Gold represented as a FixidityLib fraction
```
It this comment accurate?  It seems `value` is also expected to have been multiplied by `getDenominator()`.  Is this what we want?  Now that we're using Fixidity, could we keep the internal denominator private and unexposed?

BatiGencho replied:
param description fixed, internal denominator is private, but I have added a getter needed by the cli and contractskit

jfoutts-celo commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -121,23 +124,19 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
+    FixidityLib.Fraction memory fractionalValueReported = FixidityLib.wrap(value);
     if (rates[token].contains(msg.sender)) {
-      rates[token].update(msg.sender, value, lesserKey, greaterKey);
+      rates[token].update(msg.sender, fractionalValueReported.unwrap(), lesserKey, greaterKey);
```
Seems we're changing the type of values stored in `rates` to unwrapped Fixidity fractions?   Then we should note in the PR that the changes are not backwards compatible.

BatiGencho replied:
I have added a comment in the PR reg. that

jfoutts-celo commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -160,13 +159,21 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
+   * @notice Returns the common fixidity denominator.
+   * @return The common fixidity denominator.
+   */
+  function getDenominator() public view returns (uint256) {
```
Do we need this?

BatiGencho replied:
since we have the int. denominator as private, I have added a public method that is to be used by external sources e.g. contractskit and cli. Please let me know if any objections to that

Commit:
fixes acc. to PR comments

mcortesi commented on packages/protocol/contracts/common/MultiSig.sol:
```diff
@@ -109,15 +110,34 @@ contract MultiSig is Initializable {
     initializer
     validRequirement(_owners.length, _required)
   {
+    setInitialOwners(_owners);
```
I think the idea here is to call `addOwner` and `changeRequirement` instead of creating new setters that will only be used on initialize.

And since those functions are `external` probably makes sense to have an `_addOwner` internal function that both `initialize` and `addOwner` calls. Some for the other function

mcortesi commented on packages/protocol/contracts/governance/EpochRewards.sol:
```diff
@@ -159,6 +161,30 @@ contract EpochRewards is Ownable, Initializable, UsingPrecompiles, UsingRegistry
+   * @return True upon success.
+   * @dev unwrapped Fixidity value expected for _targetVotingYield
+   */
+  function setTargetVotingYield(uint256 _targetVotingYield) public onlyOwner returns (bool) {
```
I'm not sure for this contract, but we are changing the semantics here... since dureing the lifetime of the contract we can change the target voting yield.
I wonder if that's what we want.

 @asaj @marekolszewski  comments?

asaj replied:
+1, the scope of this ticket was more about reducing tech debt than exposing setters for values that were previously not possible to set externally

mcortesi commented on packages/protocol/contracts/governance/EpochRewards.sol:
```diff
@@ -159,6 +161,30 @@ contract EpochRewards is Ownable, Initializable, UsingPrecompiles, UsingRegistry
+   * @param _startTime The time in unix.
+   * @return True upon success.
+   */
+  function setStartTime(uint256 _startTime) public onlyOwner returns (bool) {
```
Same
 @asaj @marekolszewski  comments?

mcortesi commented on packages/protocol/contracts/governance/Governance.sol:
```diff
@@ -217,27 +219,37 @@ contract Governance is
+   * @notice Updates the last dequeue timestamp.
+   * @param _lastDequeue The last dequeue timestamp.
+   */
+  function setLastDequeue(uint256 _lastDequeue) public onlyOwner {
```
Same...
 @asaj @marekolszewski  comments?

mcortesi commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -127,6 +130,68 @@ contract StableToken is
+   * @notice Setter for the total supply.
+   * @param totalSupply The total supply to set.
+   */
+  function setTotalSupply(uint256 totalSupply) public {
```
I don't think we want a public function that lets you change the total supply

 @asaj @marekolszewski  comments?

mcortesi commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -127,6 +130,68 @@ contract StableToken is
+   * @notice Setter for the decimals of the token.
+   * @param decimals The decimals to set.
+   */
+  function setDecimals(uint8 decimals) public {
```
Same here... you can't change decimal at half of the lifecycle

mcortesi commented on packages/protocol/contracts/stability/StableToken.sol:
```diff
@@ -127,6 +130,68 @@ contract StableToken is
+   * @notice Setter for the name of the token.
+   * @param name The name to set.
+   */
+  function setName(string memory name) public {
```
What about changing `name` or `symbol` during the lifecycle?

 @asaj @marekolszewski  comments?

mcortesi reviewed (commented):
Reviewed changes for 1604

mcortesi reviewed (changes requested):
@BatiGencho where's 1597 implemented??

mcortesi commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -121,23 +124,19 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
    * @param token The address of the token for which the Celo Gold exchange rate is being reported.
-   * @param numerator The amount of tokens equal to `denominator` Celo Gold.
-   * @param denominator The amount of Celo Gold equal to `numerator` tokens.
+   * @param value The amount of tokens equal to Celo Gold represented as a FixidityLib fraction, premultiplied with the denominator
```
The idea of #1365 is not to premultiply with denominator, but instead make users report rates as a fractional number (encoded as a fixidity number).

So if rate is `1`, it should be encoded as `1000000000000000000000000`
and if it's `0.5` it should be `500000000000000000000000`

mcortesi commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -14,8 +15,10 @@ import "../common/linkedlists/SortedLinkedListWithMedian.sol";
+
   // All oracle rates are assumed to have a denominator of 2 ^ 64.
-  uint256 public constant DENOMINATOR = 0x10000000000000000;
+  uint256 private constant DENOMINATOR = 0x10000000000000000;
```
thus, this should be removed

Commit:
MultiSig -> simplify initialization with addOwner and changeRequirement

Commit:
MultiSig -> simplify initialization with addOwner and changeRequirement naming fixes,
EpochRewards -> few setters switched to privet,
Governance -> setLastDequeue() switched to private,
StableToken -> setters for name, symbol, and decimals removed

Commit:
Merge remote-tracking branch 'origin/limechain/t_1604_1597_1365' into limechain/t_1604_1597_1365

# Conflicts:
#	packages/protocol/contracts/common/MultiSig.sol

Commit:
DENOMINATOR / getDenominator removed,
numerator, denominator replaced with medianValue,
unit test update

Commit:
fixed numerator/denominatpr in soredoracles kit

Commit:
factor can be update after 18 months

asaj commented on packages/protocol/contracts/common/GasPriceMinimum.sol:
```diff
@@ -82,7 +82,11 @@ contract GasPriceMinimum is Ownable, Initializable, UsingRegistry {
       uint256 rateDenominator;
       (rateNumerator, rateDenominator) = sortedOracles.medianRate(tokenAddress);
-      return (gasPriceMinimum.mul(rateNumerator).div(rateDenominator));
+      FixidityLib.Fraction memory ratio = FixidityLib.newFixedFraction(
```
nit: s/ratio/rate

asaj commented on packages/protocol/contracts/common/MultiSig.sol:
```diff
@@ -20,6 +20,7 @@ contract MultiSig is Initializable {
   event OwnerAddition(address indexed owner);
   event OwnerRemoval(address indexed owner);
   event RequirementChange(uint256 required);
+  event RequirementSet(uint256 required);
```
Where is this used?

asaj commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -182,7 +172,7 @@ contract SortedOracles is ISortedOracles, Ownable, Initializable {
    */
   function medianRate(address token) external view returns (uint256, uint256) {
-    return (rates[token].getMedianValue(), numRates(token) == 0 ? 0 : DENOMINATOR);
+    return (rates[token].getMedianValue(), numRates(token));
```
It feels dangerous to change the behavior here unless you're 100% certain you've made changes everywhere the output of this function is consumed

asaj commented on packages/protocol/migrations/08_stabletoken.ts:
```diff
@@ -55,10 +56,10 @@ module.exports = deploymentForCoreContract<StableTokenInstance>(
       stableToken.address,
-      config.stableToken.goldPrice,
-      1,
+      new BigNumber(config.stableToken.goldPrice / 1),
```
Shouldn't we multiply by 10^24? i.e. use `toFixed` here?

asaj commented on packages/protocol/scripts/truffle/set_exchange_rate.ts:
```diff
@@ -133,7 +133,12 @@ module.exports = async (callback: (error?: any) => number) => {
-    await oracles.report(stableToken.address, numerator, denominator, lesserKey, greaterKey)
+    await oracles.report(
+      stableToken.address,
+      numerator.dividedBy(denominator),
```
Ditto, we need to use `toFixed` here

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -115,7 +117,9 @@ contract('SortedOracles', (accounts: string[]) => {
     describe('when a report has been made', () => {
       beforeEach(async () => {
-        await sortedOracles.report(aToken, 1, 1, NULL_ADDRESS, NULL_ADDRESS, { from: anOracle })
+        await sortedOracles.report(aToken, new BigNumber(1), NULL_ADDRESS, NULL_ADDRESS, {
```
I think you want `toFixed(1)` here

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -128,7 +132,7 @@ contract('SortedOracles', (accounts: string[]) => {
             const anotherOracle = accounts[i]
             await sortedOracles.addOracle(aToken, anotherOracle)
-            await sortedOracles.report(aToken, 2, 1, anOracle, NULL_ADDRESS, {
+            await sortedOracles.report(aToken, new BigNumber(2), anOracle, NULL_ADDRESS, {
```
toFixed(2)

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -170,7 +174,7 @@ contract('SortedOracles', (accounts: string[]) => {
     describe('when a report has been made', () => {
       beforeEach(async () => {
-        await sortedOracles.report(aToken, 10, 1, NULL_ADDRESS, NULL_ADDRESS, {
+        await sortedOracles.report(aToken, new BigNumber(10), NULL_ADDRESS, NULL_ADDRESS, {
```
toFixed(10)

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -253,58 +256,79 @@ contract('SortedOracles', (accounts: string[]) => {
     }
 
-    const numerator = 10
+    const numerator = 5
```
No need to define these as numerator and denominator now, you can just use a single "rate"

asaj commented on packages/protocol/test/stability/stabletoken.ts:
```diff
@@ -178,13 +179,17 @@ contract('StableToken', (accounts: string[]) => {
       beforeEach(async () => {
         await stableToken.setInflationParameters(inflationRate, SECONDS_IN_A_WEEK)
         await timeTravel(SECONDS_IN_A_WEEK, web3)
+        await timeTravel(EighteenMonthsInSeconds, web3)
```
nit: Consider time traveling the eighteen months before the wek

Commit:
MultiSig -> removed unused event, 08_stabletoken.ts and set_exchange_rate.ts the median rate is multiplied by 10^24 (toFixed), sortedoracles.ts (unittests) numerator / denomerator replaced with rate

Commit:
GasPriceMinimum namings -> rateNumerator to rate,
sortedoracles unit tests small assertion correction

Commit:
getInternalDenominator removed from SortedOracles.ts wrapper

asaj commented on packages/protocol/contracts/common/GasPriceMinimum.sol:
```diff
@@ -79,13 +79,10 @@ contract GasPriceMinimum is Ownable, Initializable, UsingRegistry {
         registry.getAddressForOrDie(SORTED_ORACLES_REGISTRY_ID)
       );
-      uint256 rateNumerator;
+      uint256 rate;
```
nit: prefer rateNumerator

asaj commented on packages/protocol/scripts/truffle/set_exchange_rate.ts:
```diff
@@ -135,7 +136,7 @@ module.exports = async (callback: (error?: any) => number) => {
     await oracles.report(
       stableToken.address,
-      numerator.dividedBy(denominator),
+      toFixed(new BigNumber(numerator).dividedBy(new BigNumber(denominator))),
```
No need to make these bignumbers, you should be able to do `toFixed(numerator / denominator)`

asaj commented on packages/protocol/test/governance/epochrewards.ts:
```diff
@@ -416,7 +416,7 @@ contract('EpochRewards', (accounts: string[]) => {
 
       it('should return one', async () => {
-        assertEqualBN(await epochRewards.getRewardsMultiplier(), toFixed(1))
+        assertEqualBN(await epochRewards.getRewardsMultiplier(), toFixed(new BigNumber(1)))
```
ditto, toFixed(1)

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -117,7 +117,7 @@ contract('SortedOracles', (accounts: string[]) => {
     describe('when a report has been made', () => {
       beforeEach(async () => {
-        await sortedOracles.report(aToken, new BigNumber(1), NULL_ADDRESS, NULL_ADDRESS, {
+        await sortedOracles.report(aToken, toFixed(new BigNumber(1)), NULL_ADDRESS, NULL_ADDRESS, {
```
ditto

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -132,7 +132,7 @@ contract('SortedOracles', (accounts: string[]) => {
             const anotherOracle = accounts[i]
             await sortedOracles.addOracle(aToken, anotherOracle)
-            await sortedOracles.report(aToken, new BigNumber(2), anOracle, NULL_ADDRESS, {
+            await sortedOracles.report(aToken, toFixed(new BigNumber(2)), anOracle, NULL_ADDRESS, {
```
ditto

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -174,7 +174,7 @@ contract('SortedOracles', (accounts: string[]) => {
     describe('when a report has been made', () => {
       beforeEach(async () => {
-        await sortedOracles.report(aToken, new BigNumber(10), NULL_ADDRESS, NULL_ADDRESS, {
+        await sortedOracles.report(aToken, toFixed(new BigNumber(10)), NULL_ADDRESS, NULL_ADDRESS, {
```
ditto

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -256,158 +256,93 @@ contract('SortedOracles', (accounts: string[]) => {
-    const numerator = 5
-    const denominator = 1
-    const expectedMedianRate = expectedMedianRateFromGiven(numerator, denominator)
+    const givenMedianRate = toFixed(new BigNumber(5).dividedBy(new BigNumber(1)))
```
ditto, `toFixed(5)`

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -256,158 +256,93 @@ contract('SortedOracles', (accounts: string[]) => {
-      assertEqualBN(actualMedianRate, expectedMedianRate)
-      assertEqualBN(numberOfRates.toNumber(), 1)
+      assertEqualBN(actualMedianRate, givenMedianRate)
+      assertEqualBN(numberOfRates, toFixed(new BigNumber(1)))
```
ditto

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -256,158 +256,93 @@ contract('SortedOracles', (accounts: string[]) => {
     describe('when there exists exactly one other report, made by this oracle', () => {
-      const newMedianRate = 12
-      const newExpectedMedianRate = expectedMedianRateFromGiven(newMedianRate, denominator)
+      const newExpectedMedianRate = toFixed(new BigNumber(12).dividedBy(new BigNumber(1)))
```
ditto `toFixed(12)`

asaj commented on packages/protocol/test/stability/sortedoracles.ts:
```diff
@@ -419,25 +354,22 @@ contract('SortedOracles', (accounts: string[]) => {
-      const anOracleExpectedMedianRate1 = expectedMedianRateFromGiven(
-        anOracleMedianRate1,
-        denominator
+      const anOracleExpectedMedianRate1 = toFixed(
```
Are these variables necessary anymore?

Commit:
unnecessary bigNumber parse removed

asaj commented:
Closing as there are open PRs that address these issues individually.
